signalfx-java/src/main/java/com/signalfx/signalflow/Channel.java

```diff
@@ -0,0 +1,97 @@
+package com.signalfx.signalflow;
+
+import java.io.Closeable;
+import java.util.Iterator;
+
+import org.apache.commons.lang3.RandomStringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Copyright (C) 2016 SignalFx, Inc. All rights reserved.
+ * 
+ * Abstract immutable representation for open channels that receive streaming data from a SignalFlow
+ * computation.
+ * 
+ * Channel objects bridge the gap between an underlying transport and a higher-level Computation
+ * object by providing a transport-agnostic and encoding-agnostic access to the stream of
+ * messages.StreamMessage objects that are received for a given computation.
+ * 
+ * Channels are iterable that return ChannelMessage instances.
+ */
+public abstract class Channel implements Iterator<ChannelMessage>, Closeable {
+
+    protected static final Logger log = LoggerFactory.getLogger(Channel.class);
+    private static final int CHANNEL_NAME_LENGTH = 8;
+
+    // should be set by subclass implementation
+    protected String type = "undefined";
```

Comment:
Take this in as a constructor parameter and have subclasses pass it in then. Right now, you're not enforcing that it gets set by the subclass implementation.